// megadraft/lib/components/MoveControl.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _icons = _interopRequireDefault(require("../icons"));

var _MegadraftBlock = _interopRequireDefault(require("./MegadraftBlock"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var Options = function Options(_ref) {
  var onClickUp = _ref.onClickUp,
      onClickDown = _ref.onClickDown,
      id = _ref.id,
      disableUp = _ref.disableUp,
      disableDown = _ref.disableDown;

  var onPointerDown = function onPointerDown(e) {
    e.preventDefault();
    e.stopPropagation();
  };

  return _react["default"].createElement("div", {
    className: "options"
  }, _react["default"].createElement("button", {
    "data-testid": "swap-up-".concat(id),
    className: "options__button options__button--up",
    onClick: onClickUp,
    disabled: disableUp,
    onPointerDown: onPointerDown
  }, _react["default"].createElement(_icons["default"].DropdownArrow, null)), _react["default"].createElement("button", {
    "data-testid": "swap-down-".concat(id),
    className: "options__button",
    onClick: onClickDown,
    disabled: disableDown,
    onPointerDown: onPointerDown
  }, _react["default"].createElement(_icons["default"].DropdownArrow, null)));
};

var Control = function Control(_ref2) {
  var children = _ref2.children,
      id = _ref2.id,
      onClickUp = _ref2.onClickUp,
      onClickDown = _ref2.onClickDown,
      isFirst = _ref2.isFirst,
      isLast = _ref2.isLast;
  return _react["default"].createElement("div", {
    className: "move-control",
    id: "move-control-".concat(id)
  }, _react["default"].createElement("div", {
    className: "move-control__target",
    "data-testid": "block-".concat(id)
  }, children), _react["default"].createElement(Options, _extends({
    id: id,
    onClickUp: onClickUp,
    onClickDown: onClickDown
  }, {
    disableUp: isFirst,
    disableDown: isLast
  })));
};

var _default = function _default(_ref3) {
  var wrapper = _ref3.wrapper,
      swapUp = _ref3.swapUp,
      swapDown = _ref3.swapDown,
      children = _ref3.children,
      isFirstBlock = _ref3.isFirstBlock,
      isLastBlock = _ref3.isLastBlock;

  var arrayChildren = _react["default"].Children.toArray(children);

  var firstChildKey = arrayChildren[0].props.children.key;
  var lastChildKey = arrayChildren[arrayChildren.length - 1].props.children.key;

  var Controlled = function Controlled(_ref4) {
    var keySwapUp = _ref4.keySwapUp,
        keySwapDown = _ref4.keySwapDown,
        children = _ref4.children;

    var onClickUp = function onClickUp() {
      return swapUp(keySwapUp);
    };

    var onClickDown = function onClickDown() {
      return swapDown(keySwapDown);
    };

    var isFirst = isFirstBlock(keySwapUp);
    var isLast = isLastBlock(keySwapDown);
    return _react["default"].createElement(_MegadraftBlock["default"], null, _react["default"].createElement(Control, _extends({
      id: keySwapUp !== keySwapDown ? "".concat(keySwapUp, "-").concat(keySwapDown) : keySwapUp
    }, {
      onClickUp: onClickUp,
      onClickDown: onClickDown,
      isFirst: isFirst,
      isLast: isLast
    }), children));
  };

  var controlledChildren = _react["default"].Children.map(children, function (child) {
    var currentKey = child.props.children.key;
    return _react["default"].createElement(Controlled, {
      keySwapUp: currentKey,
      keySwapDown: currentKey
    }, child);
  });

  return wrapper ? _react["default"].createElement(Controlled, {
    keySwapUp: firstChildKey,
    keySwapDown: lastChildKey
  }, _react["default"].cloneElement(wrapper, [], children)) : controlledChildren;
};

exports["default"] = _default;
